packages/auth/src/index.ts

```diff
@@ -255,44 +210,93 @@ export function createAuth<ListTypeInfo extends BaseListTypeInfo>({
+    if (pathname === '/signin') return;
+
+    // otherwise, redirect to signin
+    if (pathname === '/') return { kind: 'redirect', to: '/signin' };
+    return {
+      kind: 'redirect',
+      to: `/signin?from=${encodeURIComponent(req!.url!)}`,
+    };
+  }
+
+  function defaultAccessAllowed({ session }: KeystoneContext) {
+    return session !== undefined;
+  }
+
   /**
    * withAuth
    *
-   * Automatically extends config with the correct auth functionality. This is the easiest way to
-   * configure auth for keystone; you should probably use it unless you want to extend or replace
-   * the way auth is set up with custom functionality.
-   *
-   * It validates the auth config against the provided keystone config, and preserves existing
-   * config by composing existing extendGraphqlSchema functions and ui config.
+   * Automatically extends your configuration with a prescriptive implementation.
    */
   const withAuth = <TypeInfo extends BaseKeystoneTypeInfo>(
     keystoneConfig: KeystoneConfig<TypeInfo>
   ): KeystoneConfig<TypeInfo> => {
     validateConfig(keystoneConfig);
-    let ui = keystoneConfig.ui;
-    if (!keystoneConfig.ui?.isDisabled) {
+
+    let { ui } = keystoneConfig;
+    if (ui && !ui?.isDisabled) {
+      const {
+        getAdditionalFiles = [],
+        isAccessAllowed = defaultAccessAllowed,
+        pageMiddleware,
+        publicPages = [],
+      } = ui;
       ui = {
-        ...keystoneConfig.ui,
-        publicPages: [...(keystoneConfig.ui?.publicPages || []), ...publicPages],
-        getAdditionalFiles: [...(keystoneConfig.ui?.getAdditionalFiles || []), getAdditionalFiles],
-        pageMiddleware: async args =>
-          (await pageMiddleware(args)) ?? keystoneConfig?.ui?.pageMiddleware?.(args),
+        ...ui,
+        publicPages: [...publicPages, ...defaultPublicPages],
+        getAdditionalFiles: [...getAdditionalFiles, defaultGetAdditionalFiles],
+        pageMiddleware: async args => {
+          const shouldRedirect = await attemptRedirects({
+            ...args,
+            isAccessAllowed: args.isValidSession,
+            publicPages: [...publicPages, ...defaultPublicPages],
+          });
+          if (shouldRedirect) return shouldRedirect;
+          return pageMiddleware?.(args);
+        },
         isAccessAllowed: async (context: KeystoneContext) => {
-          // Allow access to the adminMeta data from the /init path to correctly render that page
-          // even if the user isn't logged in (which should always be the case if they're seeing /init)
-          const headers = context.req?.headers;
-          const host = headers ? headers['x-forwarded-host'] || headers['host'] : null;
-          const url = headers?.referer ? new URL(headers.referer) : undefined;
-          const accessingInitPage =
-            url?.pathname === '/init' &&
-            url?.host === host &&
-            (await context.sudo().query[listKey].count({})) === 0;
```

Comment:
I replaced this with
```
if (pathname === '/__nextjs_original-stack-frame') return
```

And for the meta endpoint the replacement for `isAccessAllowed` is simply to return `true` for this when `hasInitFirstItemConditions`.

Reply:
Now replaced with https://github.com/keystonejs/keystone/blob/f6fb299cb987a27ce9c0885620dc2b68a6186ef8/packages/core/src/lib/server/createAdminUIMiddleware.ts#L46